Update of show1D to accept numpy arrays and change subplot behaviour.

gfardell opened the pull request:
## Changes
Update of show1D.
- Previously had buggy behaviour with 3D data. None of the examples in this PR worked.
- Each slice is now plotted on a new subplot with a generated title and axis labels
- If multiple datasets are listed then they same slices are applied to each and they are added to the slice subplot. This also adds a legend.
- If slice is not provided then centre vector in each dimension will be shown on N-dim subplots
- Works with numpy ndarrays

Some changes in argument naming breaking backward compatibility but adding clarity.

New behaviour:
```
data = dataexample.SIMULATED_SPHERE_VOLUME.get()
show1D(data, slice_list=None)
```
![image](https://github.com/user-attachments/assets/b4767130-2bf5-4b84-ac7d-7ebdc91175da)

```
data = dataexample.SIMULATED_SPHERE_VOLUME.get()
show1D(data, slice_list=[("Horizontal_x", 30),("Horizontal_y", 40)])
```
![image](https://github.com/user-attachments/assets/7a0e68b3-de01-4dcf-a6ec-166190da374a)

```
data = dataexample.SIMULATED_SPHERE_VOLUME.get()
show1D(data, slice_list=[[("Vertical",62),("Horizontal_y",70)],[("Vertical",61),("Horizontal_y",70)],[("Vertical",67),("Horizontal_x",70)]])
```
![image](https://github.com/user-attachments/assets/b6c62e66-1723-4b3e-a4a8-8d3f753ab5f7)


```
data = dataexample.SIMULATED_SPHERE_VOLUME.get()
data2 = data*0.9
show1D([data, data2], slice_list=[("Horizontal_x", 30),("Horizontal_y", 40)])
```
![image](https://github.com/user-attachments/assets/6f12d187-0f2e-4d02-bd13-2564ee534c90)

```
data = dataexample.SIMULATED_SPHERE_VOLUME.get()
data2 = data*0.9
show1D([data.array, data2], slice_list=[(1, 30),(2, 40)])
```
data = dataexample.SIMULATED_SPHERE_VOLUME.get()
data2 = data*0.9
show1D([data.array, data2], slice_list=[(1, 30),(2, 40)])



Comparing with previous behaviour (from documentation examples):
```
data = PEPPERS.get()
data_channel0 = data.get_slice(channel=0)
data_channel1 = data.get_slice(channel=1)
show1D([data_channel0, data_channel1], slice_list=[('horizontal_x', 256)],label=['Channel 0', 'Channel 1'], line_styles=["--", "-"])
```
Previous:
![image](https://github.com/user-attachments/assets/b2221a35-87b5-436f-a76b-00a2073faf9c)
New:
![image](https://github.com/user-attachments/assets/070eb6e8-251f-4214-9474-5c63ea05d01b)


```
data = PEPPERS.get()
slices = [[('channel', 0), ('horizontal_x', 256)], [('channel', 1), ('horizontal_y', 256)]]
show1D(data, slice_list=slices, title=['Channel 0', 'Channel 1'])
```
Previous:
![image](https://github.com/user-attachments/assets/4e9255fe-d171-495c-a42d-d622896d75e7)
New:
![image](https://github.com/user-attachments/assets/6ec71edd-ee2f-4c1e-b005-2f68f4c166a1)

Or with default titles:
![image](https://github.com/user-attachments/assets/381efdc1-fdbe-43ec-a00d-cccf42718779)


## Testing you performed
> Please add any demo scripts to https://github.com/TomographicImaging/CIL-Demos/tree/main/misc


## Related issues/links


## Checklist

- [x] I have performed a self-review of my code
- [x] I have added docstrings in line with the guidance in the developer guide
- [x] I have updated the relevant documentation
- [x] I have implemented unit tests that cover any new or modified functionality
- [x] CHANGELOG.md has been updated with any functionality change
- [x] Request review from all relevant developers
- [x] Change pull request label to 'Waiting for review'

## Contribution Notes

Please read and adhere to the [developer guide](https://tomographicimaging.github.io/CIL/nightly/developer_guide/) and local patterns and conventions.

- [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in CIL (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License
- [x] I confirm that the contribution does not violate any intellectual property rights of third parties

gfardell commented:
Failing CI in unrelated matplotlib call. Not on the actions that are focused on the tests (which is odd)

<img width="1016" alt="image" src="https://github.com/user-attachments/assets/2bb8e29c-663c-4525-ae22-1f5f2274b357" />

lauramurgatroyd commented on Wrappers/Python/cil/utilities/display.py:
```diff
@@ -103,277 +103,274 @@ def save(self,filename, **kwargs):
-    Attributes
-    ----------
-    figure : matplotlib.figure.Figure
+        Overides the default line style(s) for each line plot. Must be a list of styles equal to the number of datasets.
```
```suggestion
        Overrides the default line style(s) for each line plot. Must be a list of styles equal to the number of datasets.
```

lauramurgatroyd commented on Wrappers/Python/cil/utilities/display.py:
```diff
@@ -103,277 +103,274 @@ def save(self,filename, **kwargs):
-    figure : matplotlib.figure.Figure
+        Overides the default line style(s) for each line plot. Must be a list of styles equal to the number of datasets.
+    axis_labels : tuple of str, list of tuples, optional
+        Overides the default axis labels in the form (x_axis_label, y_axis_label). Must be a single tuple or a list of tuples equal to the number of plots.
```
```suggestion
        Overrides the default axis labels in the form (x_axis_label, y_axis_label). Must be a single tuple or a list of tuples equal to the number of plots.
```

lauramurgatroyd commented on Wrappers/Python/cil/utilities/display.py:
```diff
@@ -103,277 +103,274 @@ def save(self,filename, **kwargs):
+        Overides the default line style(s) for each line plot. Must be a list of styles equal to the number of datasets.
+    axis_labels : tuple of str, list of tuples, optional
+        Overides the default axis labels in the form (x_axis_label, y_axis_label). Must be a single tuple or a list of tuples equal to the number of plots.
+    plot_size : tuple, default=(8,5)
```
In the init this is called size and defaults to (8,3)

lauramurgatroyd reviewed (changes requested):
This is great, I like the new functionality and lots of examples?

Should it go in v25.0.0 as it breaks backwards compatability?
It would be good to have full details of compatability break in the change log.

I checked the docs render and it looks good

lauramurgatroyd commented:
> Failing CI in unrelated matplotlib call. Not on the actions that are focused on the tests (which is odd)
> <img alt="image" width="1016" src="https://private-user-images.githubusercontent.com/47746591/399719024-2bb8e29c-663c-4525-ae22-1f5f2274b357.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4Mjk3MzAsIm5iZiI6MTczNTgyOTQzMCwicGF0aCI6Ii80Nzc0NjU5MS8zOTk3MTkwMjQtMmJiOGUyOWMtNjYzYy00NTI1LWFlMjItMWY1ZjIyNzRiMzU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTAyVDE0NTAzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwY2E3N2I3YmVlMzM4YzhjMDI1MDY2ZTA0N2E4NjZlZDQ5Mzg2ZmM5MGY4ZDNmMzU3MzNmYTRmNzhiMjIzMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.QRbK58GlbSSDuu7cQLm2HEon7AeV-boTR8aE5AFpl7I">

This test is skipped if TIGRE or IPP are not installed - could it be being skipped by the other actions?

gfardell commented:
I have reverted the argument to `title` as this is consistent with show2D (it will still take a list of titles as stated by the docstring). Updated changelog.

lauramurgatroyd commented on Wrappers/Python/cil/utilities/display.py:
```diff
@@ -103,277 +103,276 @@ def save(self,filename, **kwargs):
+        Each slice should be a list of tuples, where each tuple is a (axis, index). Each axis can be described as a string or integer representing the axis to slice along.
+        To plot multiple slices, pass a list of slices.        
+    dataset_labels : list of str, optional
+        Overides the default legend labels for each dataset displayed in the plot legend. Must be a list of strings equal to the number of datasets.
```
```suggestion
        Overrides the default legend labels for each dataset displayed in the plot legend. Must be a list of strings equal to the number of datasets.
```

Sorry, missed one! maybe do a find and replace on the file?

lauramurgatroyd reviewed (approved):
Thanks for the updates!

hrobarts commented on Wrappers/Python/cil/utilities/display.py:
```diff
@@ -103,277 +103,276 @@ def save(self,filename, **kwargs):
-    ----------
-    figure : matplotlib.figure.Figure
 
+    Returns
```
Maybe remove this because the init doesn't return?

hrobarts reviewed (approved):
This looks like a good update! I just have a couple of small comments
